sunnypilot/selfdrive/controls/lib/nnlc/nnlc.py

```diff
@@ -0,0 +1,262 @@
+    past_rolls = [self.roll_deque[min(len(self.roll_deque) - 1, i)] for i in self.history_frame_offsets]
+    future_rolls = [roll_pitch_adjust(np.interp(t, ModelConstants.T_IDXS, self.model_v2.orientation.x) + roll,
+                                      np.interp(t, ModelConstants.T_IDXS, self.model_v2.orientation.y) + self.pitch_last) for t in
+                    adjusted_future_times]
+    past_lateral_accels_desired = [self.lateral_accel_desired_deque[min(len(self.lateral_accel_desired_deque) - 1, i)]
+                                   for i in self.history_frame_offsets]
+    future_planned_lateral_accels = [np.interp(t, ModelConstants.T_IDXS, self.model_v2.acceleration.y) for t in
+                                     adjusted_future_times]
+
+    # compute NNFF error response
+    nnff_setpoint_input = [CS.vEgo, setpoint, self.lateral_jerk_setpoint, roll] \
+                          + [setpoint] * self.past_future_len \
+                          + past_rolls + future_rolls
+    # past lateral accel error shouldn't count, so use past desired like the setpoint input
+    nnff_measurement_input = [CS.vEgo, measurement, self.lateral_jerk_measurement, roll] \
+                             + [measurement] * self.past_future_len \
+                             + past_rolls + future_rolls
+    torque_from_setpoint = self.flux_model.evaluate(nnff_setpoint_input)
+    torque_from_measurement = self.flux_model.evaluate(nnff_measurement_input)
+    self._pid_log.error = torque_from_setpoint - torque_from_measurement
+
+    # compute feedforward (same as nn setpoint output)
+    error = setpoint - measurement
+    friction_input = self.lat_accel_friction_factor * error + self.lat_jerk_friction_factor * self.lookahead_lateral_jerk
+    nn_input = [CS.vEgo, desired_lateral_accel, friction_input, roll] \
+               + past_lateral_accels_desired + future_planned_lateral_accels \
+               + past_rolls + future_rolls
+    self._ff = self.flux_model.evaluate(nn_input)
+
+    # apply friction override for cars with low NN friction response
+    if self.flux_model.friction_override:
+      self._pid_log.error += self.torque_from_lateral_accel(LatControlInputs(0.0, 0.0, CS.vEgo, CS.aEgo), self.torque_params,
+                                                      friction_input,
+                                                      lateral_accel_deadzone, friction_compensation=True,
+                                                      gravity_adjusted=False)
+
+  def update_stock_lateral_jerk(self, CS, setpoint, measurement, roll_compensation, desired_lateral_accel, actual_lateral_accel,
+                                lateral_accel_deadzone, gravity_adjusted_lateral_accel):
+    if (self.enabled and self.model_valid) or not self.use_lateral_jerk:
+      return
+
+    _error = desired_lateral_accel - actual_lateral_accel
+    friction_input = self.lat_accel_friction_factor * _error + self.lat_jerk_friction_factor * self.actual_lateral_jerk
```

Comment:
Here, it should be the look ahead lateral jerk (desired) that's used, not actual lateral jerk.

Reply:
Perhaps a better approach, that avoids redundant code, is to call this method from inside the update feedforward method. That way, look ahead lateral jerk can be computed once and passed to this method.

Reply:
Or save the lookahead jerk to an attribute in the feedforward method so it can be used in the lateral jerk method.